File: nvim/.config/nvim/lua/ppm/toolkit/frontmatter-gemini.lua
-- frontmatter.lua

local Frontmatter = {}

--- Trims leading and trailing whitespace from a string.
local function trim(str)
  return str:gsub("^%s*", ""):gsub("%s*$", "")
end

--- Parses a YAML value string into its Lua representation.
local function parse_value(value_string)
  local value = trim(value_string)
  if value:match("^%[%s*%]$") then return {} end
  if value:match("^%{%s*%}$") then return {} end
  if value:match("^%d+$") then return tonumber(value) end
  if value:match("^%d+%.%d+$") then return tonumber(value) end
  if value == "true" then return true end
  if value == "false" then return false end

  return value
end


--- Parses a YAML frontmatter string.
function Frontmatter.parse(frontmatter_string)
  local data = {}
  local lines = frontmatter_string:gmatch("[^\r\n]+")()

  local in_frontmatter = false
  local current_key = nil
  local in_list = false

  for _, line in ipairs(lines) do
    line = trim(line)

    if line == "---" then
      if not in_frontmatter then
        in_frontmatter = true
      else
        in_frontmatter = false
        break
      end
    elseif in_frontmatter then
      local key_match = line:match("([^:]+):")

      if key_match then
        current_key = trim(key_match)
        data[current_key] = data[current_key] or {}
        in_list = false
      elseif line:match("^%s*-%s") then -- List item
        if current_key then
          local value_string = line:match("^%s*-%s*(.+)")
          local value = parse_value(value_string)

          if type(data[current_key]) == "table" and not in_list then
            data[current_key] = { value }
            in_list = true
          elseif type(data[current_key]) == "table" and in_list then
            table.insert(data[current_key], value)
          else
            return nil, "Invalid list syntax"
          end
        else
          return nil, "List item without a key"
        end
      elseif current_key and not in_list and line ~= "" then -- Single-line value
        local value = parse_value(line)

        data[current_key] = value
      end
    end
  end

  if not in_frontmatter then
    return nil, "Frontmatter not found or malformed"
  end

  return data, nil
end

return Frontmatter

File: nvim/.config/nvim/lua/ppm/toolkit/frontmatter-gemini.ts
import * as O from 'fp-ts/Option';
import * as A from 'fp-ts/Array';
import * as E from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';

type FrontmatterData = Record<string, string | number | boolean | {} | string[]>;

const trim = (str: string): string => str.trim();

const parse_value = (valueString: string): string | number | boolean | {} | string[] => {
  const value = trim(valueString);
  if (value === "[]") return [];
  if (value === "{}") return {};
  const num = Number(value);
  if (!isNaN(num)) return num;
  if (value === "true") return true;
  if (value === "false") return false;
  return value;
};

const parse_line = (line: string): O.Option<[string, string]> => {
  const match = line.match(/^([^:]+):\s*(.+)$/);
  return match ? O.some([trim(match[1]), match[2]]) : O.none;
};

const parse_list_item = (line: string): O.Option<string> => {
  const match = line.match(/^\s*-\s*(.+)$/);
  return match ? O.some(match[1]) : O.none;
};

const handle_line_with_value = (data: FrontmatterData, key: string, valueString: string): E.Either<string, FrontmatterData> => {
  const parsedValue = parse_value(valueString)
  return E.right({ ...data, [key]: parsedValue })
}

const handle_list_item = (data: FrontmatterData, key: string, item: string): E.Either<string, FrontmatterData> => {
  const parsedItem = parse_value(item)
  const existingValue = data[key]
  if (Array.isArray(existingValue)) {
    return E.right({ ...data, [key]: [...existingValue, parsedItem] })
  } else if (existingValue === undefined) {
    return E.right({ ...data, [key]: [parsedItem] })
  } else {
    return E.left("Invalid list syntax")
  }
}

const process_line = (
  data: FrontmatterData,
  currentKey: O.Option<string>,
  line: string
): E.Either<string, FrontmatterData> => {

  const parsedLine = parse_line(line)
  if (O.isSome(parsedLine)) {
    const [key, valueString] = parsedLine.value
    return handle_line_with_value(data, key, valueString)
  }

  const parsedListItem = parse_list_item(line)
  if (O.isSome(parsedListItem)) {
    const item = parsedListItem.value
    return pipe(
      currentKey,
      O.fold(
        () => E.left("List item without a key"),
        (key) => handle_list_item(data, key, item)
      )
    )
  }

  return E.right(data);
};

export const parseFrontmatter = (frontmatterString: string): E.Either<string, FrontmatterData> => {
  const lines = frontmatterString.split('\n').map(trim);

  const initialData: FrontmatterData = {};

  let inFrontmatter = false;
  let currentKey: O.Option<string> = O.none;

  for (const line of lines) {
    if (line === "---") {
      inFrontmatter = !inFrontmatter;
      if (!inFrontmatter) break; // Exit after closing "---"
      continue; // Skip the "---" line itself
    }

    if (inFrontmatter) {
      const result = process_line(initialData, currentKey, line)
      if (E.isLeft(result)) {
        return result
      } else {
          initialData = result.right
          currentKey = parse_line(line).map(([k, _]) => k)
      }
    }
  }

  return inFrontmatter ? E.left("Frontmatter not closed") : E.right(initialData);
};